Improve method discovery and filtering performance

lahma opened the pull request:
First of all, sorry for not pinging beforehand and discussing with you about the change. I got a bit carried away while profiling and worked with the fixes as I went on.

## TLDR;

80k parametrized test case suite (loading suite, counting test cases) down from almost 11 minutes to less than one second when using Rider. So for this specific case it's __over 700x improvement__.

## Backstory

I'm generating test source code for running [ECMAScript test suite](https://github.com/tc39/test262). This test suite is heavy with 86k+ test cases (about 40k files which are tested in both strict and non-strict mode) and they are ideal use case for using `TestCaseAttribute` defining the files for each method (where method is usually a JavaScript target like `Array.prototype.filter`, having many test cases, e.g. files). 

## Problem

I got the generation working and then started to run them with Rider... or actually I didn't as the whole unit testing sub-system froze when I tried to run the full test suite. I could run small subsets without the IDE's unit testing session hanging in waiting state. The pending to start state before tests would actually run could take minutes. Seems that Rider has different approach than VS (based on earlier issue it sends more like full names) to sending tests - Rider uses OR with all the ids - 80k ids that is. Which will cause the inclusion check algorithm to loop with a large set of data using O(n^2) .

[Here's example of generated test case](https://gist.github.com/lahma/ec546e690aebfed4d186f3a02c80b037) and how I then later profiled NUnit with benchmark test case run in console, using data from Rider invocation (`FrameworkController` parameters). I tweaked it a bit on the way, but results are calculated using the same code.

It seems that it's enough to have a test case such as below to make Rider freeze with unit testing operations. This should mimic the situation quite well be creating the 86 test cases we are testing here.

```csharp
public class TestCaseTest
{
    [Test]
    public void Test([Range(1, 86_000)] int i)
    {
    }
}
```


## Results before

Processing 86607 test cases. Loading them and then discovery without filter and then counting with id filter containing every id found. Ryzen 5950X with 32GB RAM. Almost 11 minutes to load tests and go trough them using a OR'ed filter. ExploreTests is using empty filter in Rider's call chain.

```
LoadTests took: 00:00:27.5665738 // total 86k tests, mostly from TestCase attributes
ExploreTests took: 00:00:00.0037093 // empty filter
CountTests took: 00:10:09.3055713 // 86k OR Id filter items - yes, 10 MINUTES
```

## Fixes

Fixes consist of two parts:

### Loading tests: reducing memory allocations and re-calculations when building test cases

It's very costly to do reflection and inspecting same targets multiple times. For a method having 20 test case attributes all introspection against the test method would be done 20 times, when only parameters for the test case in construction will vary.

Now caching static information about `IMethodInfo` for subsequent accesses.

### Filtering tests: combine `OR` filter with same-kind, non-regexp items to single hash-based InFilter

This builds on the earlier https://github.com/nunit/nunit/issues/3786 and its PR https://github.com/nunit/nunit/pull/3787 . Now detecting more filter types and target fields and creating efficient hash set for that case. Created new `InFilter` type dedicated to matching.

/cc @pakrym 

## Results after

```
LoadTests took: 00:00:00.6497215 // total 86k tests, mostly from TestCase attributes
ExploreTests took: 00:00:00.0030948 // empty filter
CountTests took: 00:00:00.1815572 // 86k OR Id filter items
```

/cc @AndreyAkinshin for Rider performance in Unit testing (VS has different filter logic for full test suite I guess)

Related issues: https://github.com/nunit/nunit/issues/3601 https://github.com/nunit/nunit/issues/3786 https://github.com/nunit/nunit/issues/4009 https://github.com/nunit/nunit3-vs-adapter/issues/660

ChrisMaddock commented on src/NUnitFramework/framework/Internal/Builders/MethodInfoCache.cs:
```diff
@@ -0,0 +1,62 @@
+        }
+
+        /// <summary>
+        /// Memoization of TestMethod information to reduce subsequent allocations from parameter and attribute information.
```
```suggestion
        /// Memorization of TestMethod information to reduce subsequent allocations from parameter and attribute information.
```

lahma replied:
I think I got it right https://en.wikipedia.org/wiki/Memoization , but I can change if you want to 🙂

ChrisMaddock replied:
Oops, sorry! New word to me, and my browsers spellchecker! 😆

ChrisMaddock reviewed (commented):
The performance numbers intrigued me on this one so I had a look through. This looks brilliant!

I couldn't find any problem with the code here, and the numbers are really impressive. The only thing that jumped out to me was where this would benefit from some additional testing around the new logic in `TestFilter.FromXml` for when an InFilter should and shouldn't be built - more to protect against future code changes in this area. (I wonder if it may also be worth breaking that logic out into a separate method?)

Looks like a great improvement @lahma! 👍

lahma commented:
I'll improve test coverage and check if breaking into separate method is easy.

siprbaum commented on src/NUnitFramework/framework/Internal/AttributeProviderWrapper.cs:
```diff
@@ -52,14 +51,27 @@ public object[] GetCustomAttributes(bool inherit)
         {
-            return attributes
-                   .OfType<T>()
-                   .ToArray();
+            List<T> filtered = null;
+            foreach (var attribute in attributes)
+            {
+                if (attribute is T t)
+                {
+                    filtered ??= new List<T>();
+                    filtered.Add(t);
+                }
+            }
+
+            return filtered?.ToArray() ??
+#if NETSTANDARD2_0_OR_GREATER
+                Array.Empty<T>();
+#else
+                new T[0];
+#endif
```
When glancing over the code I really got confused  about this change, because from a readibility point of view it is definitely worse than before 😞 Took me quite a few brain cycles to see that it just replaces the linq usage and that it is correct (which it is, as far as I can tell!).

So at least I would have expected a seperate commit doing this "get rid of linq" change,  accompanied with the motivating performance numbers in the commit message telling WHY this  decrease in readability is needed, e.g. because it improves the performance. Would make it easier to swallow that bitter pill here ;-)
Right now it is embedded in the "Cache method metadata" commit, which this change above has nothing to do with.

Please take this comment with a grain of salt, as this is just a "drive by review", as I am not a contributer to this project and therefore I don't think my opinion really matters here...

lahma replied:
Fair enough, optimizations tend to get a bit verbose and harder to read. "Cache method metadata" might not tell the store, but I put it under the "Reduce allocations" sub-heading under PR 😉 So it was one of the things that contributed to performance improvements, I should have included the rough allocation numbers in hindsight though.

This method was called quite a lot and it usually returned empty array at the end.  It was hot w.r.t to allocations so it was part of the cleanup for the biggest offenders. I might create a separate PR for other long-hanging fruit but this already got a bit out of hand... 

Thanks for the drive-by, all feedback is always appreciated!

lahma commented:
@ChrisMaddock Thank you for the review, I've addressed your feedback in latest commit (except for the grammar), by moving optimization step into `InFilter` nullables were introduced so I had to tweak signatures a bit.

ChrisMaddock commented:
Nice refactor! 👍 That structure looks all a bit cleaner to me, I like it!

lahma commented:
So it's been a month, all horrible things going on aside, is there something I can do to push this forward?

stevenaw commented:
Sorry it's taken so long to get to this @lahma 
We usually like to have two sets of reviews for changes. I hope to be able to get to this myself this weekend or sometime next week

mikkelbu commented on src/NUnitFramework/framework/Internal/Filters/InFilter.cs:
```diff
@@ -0,0 +1,100 @@
+                var filter = orFilter.Filters[i];
+
+                // make sure invariants are valid
+                commonType ??= filter.GetType();
```
Would it make any difference if we do this initialization outside the loop ? E.g. remove this line and change the declaration to
```
Type? commonType = hashCommonTypeAndNonRegex ? orFilter.Filters[0] : null;
```

lahma replied:
That would change the semantics, the next line ensure that every filter in list has the same type, which is a requirement.

`hashCommonTypeAndNonRegex &= commonType == filter.GetType();`

mikkelbu replied:
Isn't the semantics the same?

In both cases `commonType` will be the type of the first element in `orFilter.Filters`. I was just wondering if it would make any difference to set `commonType` outside the loop rather than in each iteration checking if `commonType` is `null` and in that case assign it to the type of the current filter (meaning that this will only be set in the first iteration, but we will check in each iteration)

mikkelbu replied:
This was mostly a question as I was curious whether it made any difference (my feeling is that it don't, so I don't think this should hold back this PR). Thanks for the great work @lahma :+1 and thanks for having patience with us :)

mikkelbu commented on src/NUnitFramework/framework/Internal/Filters/InFilter.cs:
```diff
@@ -0,0 +1,100 @@
+
+        public override TNode AddToXml(TNode parentNode, bool recursive)
+        {
+            throw new NotImplementedException();
```
I think this will give us problems in runners that also write the filters in the XML (e.g. NUnitLite, the console etc). I guess we could write the filter as a `OrFilter` (https://docs.nunit.org/articles/nunit/technical-notes/usage/Test-Filters.html), but then we need to store the original OrFilter also (or at least have more information).

lahma replied:
My thinking was that this is dynamic, e.g. translated runtime so it would not matter. I probably don't understand the full ramifications, should this be addressed in this PR, will this be breaking things?

mikkelbu commented on src/NUnitFramework/tests/Internal/Filters/InFilterTests.cs:
```diff
@@ -0,0 +1,83 @@
+// ***********************************************************************
+// Copyright (c) 2015 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
```
```suggestion
// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
```

mikkelbu reviewed (commented):
This looks very promissing. I think we need to handle the `NotImplementedException` in InFilter.cs. The rest of the comments are minor issues and questions/comments.

Great work @lahma

stevenaw commented on src/NUnitFramework/framework/Internal/AttributeProviderWrapper.cs:
```diff
@@ -23,12 +23,11 @@
 namespace NUnit.Framework.Internal
 {
-    internal class AttributeProviderWrapper<T> : ICustomAttributeProvider
+    internal sealed class AttributeProviderWrapper<T> : ICustomAttributeProvider
```
Thanks for this 😄

lahma replied:
Ideally all private/internal things should be sealed, it allows CLR some more room to optimize. This is what the coreclr/runtime did, made everything sealed they could. Might be a micro-optimization, but can be important to things like comparers/equality checkers.

stevenaw commented on src/NUnitFramework/framework/Internal/Filters/CompositeFilter.cs:
```diff
@@ -21,7 +21,6 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 // ***********************************************************************
 
-using System;
```
Thanks for doing opportunistic cleanup like this too

lahma replied:
I try to do this only for files I touch, this was a bit on the fence as the addition of `sealed` was not required for the functionality 🤷🏻‍♂️

stevenaw reviewed (approved):
This looks great thanks! I asked a few questions but no real additional comments or concerns beyond @mikkelbu 's. Nice job.

lahma commented on src/NUnitFramework/tests/Internal/Filters/OrFilterTests.cs:
```diff
@@ -241,7 +241,7 @@ public void BuildFromXmlFullName()
                 $"<filter><or><test>{DUMMY_CLASS}</test><test>{ANOTHER_CLASS}</test></or></filter>");
 
-            Assert.That(filter, Is.TypeOf<OrFilter>());
+            Assert.That(filter, Is.TypeOf<InFilter>());
```
I moved this to `InFilterTests.cs`, not sure about create `OrFilter` part, it's `or` in XML so test case makes sure it's transformed?

mikkelbu replied:
I think this is fine. The second bullet was a mistake on my part. I just wanted to be sure we had tests of XML => OrFilter, but this is covered by `BuildFromXml` and `BuildFromXmlFullNameRegex`.

lahma commented on src/NUnitFramework/framework/Internal/Builders/MethodInfoCache.cs:
```diff
@@ -0,0 +1,76 @@
+                IsAsyncOperation = AsyncToSyncAdapter.IsAsyncOperation(method.MethodInfo);
+                IsVoidOrUnit = Reflect.IsVoidOrUnit(method.ReturnType.Type);
+
+                // TODO could probably go trough inherited and non inherited in two passes instead of multiple
```
I think it could be a separate PR, the performance is the same compared to current, this is a not that current code (even before the PR) loops same collections multiple times, which is unnecessary, but it would complicate this PR, I think.

lahma commented on src/NUnitFramework/framework/Internal/AttributeProviderWrapper.cs:
```diff
@@ -52,14 +51,27 @@ public object[] GetCustomAttributes(bool inherit)
         }
 
-        private static T[] GetFiltered(IEnumerable<object> attributes)
+        private static T[] GetFiltered(object[] attributes)
```
Yes it's always a bit bad for perf to hide an array behind an interface. They never really are interface-based so a slower wrapper will be introduced. CLR is highly optimized when it sees an actual array. Not a matter when called a couple of times but when in thousands, it adds up.

lahma commented:
@stevenaw thank you for your review. I responded to comments, tried to improve the code based on suggestion, but leaving the question about `AddToXml` open. It also seems that CI is now broken due to using some obsolete build infrastructure 😬

stevenaw commented:
Thanks @lahma 
The build should now be fixed of you merge latest into your branch. I'm less familiar with the implications of the xml line of questions, so I think Mikkel would know better about that

lahma commented:
@stevenaw @mikkelbu I've rebased this branch against `v3.13-dev` and implemented the `AddToXml` including test coverage. Now `AddToXml` generates same XML as the original `OrFilter` that was transformed into `InFilter`.

mikkelbu commented:
Thanks for the work @lahma :+1. I'll be merging it now.

lahma commented:
Thanks to everyone involved and helping out iterating the final version!
